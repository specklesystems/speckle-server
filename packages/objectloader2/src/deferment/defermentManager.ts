import { DeferredBase } from './deferredBase.js'
import { CustomLogger } from '../types/functions.js'
import { Item, Base } from '../types/types.js'
import { BaseCache } from './BaseCache.js'

export class DefermentManager {
  private deferments: Map<string, DeferredBase> = new Map()
  private logger: CustomLogger
  private disposed = false
  private cache: BaseCache

  constructor(cache: BaseCache, logger: CustomLogger) {
    this.cache = cache
    this.logger = logger
    }

    private now(): number {
      return Date.now()
    }

  get(id: string): DeferredBase | undefined {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    return this.deferments.get(id)
  }

  defer(params: { id: string }): [Promise<Base>, boolean] {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    const item = this.cache.get(params.id)
    if (item) {
      return [Promise.resolve(item.base!), true]
    }
    const deferredBase = this.deferments.get(params.id)
    if (deferredBase) {
      return [deferredBase.getPromise(), true]
    }
    const notYetFound = new DeferredBase(
      params.id
    )
    this.deferments.set(params.id, notYetFound)
    return [notYetFound.getPromise(), false]
  }

  undefer(item: Item, requestItem: (id: string) => void): void {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    const base = item.base
    if (!base) {
      this.logger('undefer called with no base', item)
      return
    }
    this.cache.add(item, (id) => {
      if (!this.deferments.has(id)) {
        requestItem(id)
      }
    })

    //order matters here with found before undefer
    const deferredBase = this.deferments.get(item.baseId)
    if (deferredBase) {
      deferredBase.found(base, item.size || 0)
      this.deferments.delete(item.baseId)
    }
  }

  dispose(): void {
    if (this.disposed) return
    this.disposed = true
    this.clearDeferments()
  }

  private clearDeferments(): void {
    let waiting = 0
    for (const deferredBase of this.deferments.values()) {

      if (deferredBase.getBase() === undefined) {
        waiting++
      }
    }
    this.deferments.clear()
    this.logger('cleared deferments, left', waiting)
  }
}
