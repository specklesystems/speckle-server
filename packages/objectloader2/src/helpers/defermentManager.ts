import { DeferredBase } from './deferredBase.js'
import { Base, CustomLogger, Item } from '../types/types.js'
import { DefermentManagerOptions } from '../operations/options.js'

export class DefermentManager {
  private deferments: Map<string, DeferredBase> = new Map()
  private timer?: ReturnType<typeof setTimeout>
  private logger: CustomLogger
  private currentSize = 0
  private disposed = false

  constructor(private options: DefermentManagerOptions) {
    this.resetGlobalTimer()
    this.logger = options.logger || ((): void => {})
  }

  private now(): number {
    return Date.now()
  }

  isDeferred(id: string): boolean {
    return this.deferments.has(id)
  }

  get(id: string): DeferredBase | undefined {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    return this.deferments.get(id)
  }

  async defer(params: { id: string }): Promise<Base> {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    const now = this.now()
    const deferredBase = this.deferments.get(params.id)
    if (deferredBase) {
      deferredBase.setAccess(now)
      return deferredBase.getPromise()
    }
    const notYetFound = new DeferredBase(
      this.options.ttlms,
      params.id,
      now + this.options.ttlms
    )
    this.deferments.set(params.id, notYetFound)
    return notYetFound.getPromise()
  }

  undefer(item: Item): void {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    const now = this.now()
    this.currentSize += item.size || 0
    //order matters here with found before undefer
    const deferredBase = this.deferments.get(item.baseId)
    if (deferredBase) {
      deferredBase.found(item)
      deferredBase.setAccess(now)
    } else {
      const existing = new DeferredBase(this.options.ttlms, item.baseId, now)
      existing.found(item)
      this.deferments.set(item.baseId, existing)
    }
  }

  private resetGlobalTimer(): void {
    const run = (): void => {
      this.cleanDeferments()
      this.timer = setTimeout(run, this.options.ttlms)
    }
    this.timer = setTimeout(run, this.options.ttlms)
  }

  dispose(): void {
    if (this.disposed) return
    this.disposed = true
    if (this.timer) {
      clearTimeout(this.timer)
      this.timer = undefined
    }
    this.clearDeferments()
  }

  private clearDeferments(): void {
    let waiting = 0
    for (const deferredBase of this.deferments.values()) {
      deferredBase.done(0)
      if (deferredBase.getItem() === undefined) {
        waiting++
      }
    }
    this.currentSize = 0
    this.deferments.clear()
    this.logger('cleared deferments, left', waiting)
  }

  private cleanDeferments(): void {
    const maxSizeBytes = this.options.maxSizeInMb * 1024 * 1024
    if (this.currentSize < maxSizeBytes) {
      this.logger(
        'deferments size is ok, no need to clean',
        this.currentSize,
        maxSizeBytes
      )
      return
    }
    const now = this.now()
    let cleaned = 0
    const start = performance.now()
    for (const deferredBase of Array.from(this.deferments.values())
      .filter((x) => x.isExpired(now))
      .sort((a, b) => this.compareMaybeBasesBySize(a.getItem(), b.getItem()))) {
      if (deferredBase.done(now)) {
        this.currentSize -= deferredBase.getItem()?.size || 0
        this.deferments.delete(deferredBase.getId())
        cleaned++
        if (this.currentSize < maxSizeBytes) {
          break
        }
      }
    }
    this.logger(
      'cleaned deferments, cleaned, left',
      cleaned,
      this.deferments.size,
      performance.now() - start
    )
    return
  }

  compareMaybeBasesBySize(a: Item | undefined, b: Item | undefined): number {
    if (a === undefined && b === undefined) return 0
    if (a === undefined) return -1
    if (b === undefined) return 1
    return this.compareMaybe(a.size, b.size)
  }

  compareMaybe(a: number | undefined, b: number | undefined): number {
    if (a === undefined && b === undefined) return 0
    if (a === undefined) return -1
    if (b === undefined) return 1
    return a - b
  }
}
