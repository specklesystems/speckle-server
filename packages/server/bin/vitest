#!/usr/bin/env node
import path from 'path'

/**
 * Find bin and run it (we don't want to hardcode a specific node_modules path).
 * We use this so that we can pass in specific flags to node before the bin even begins to run
 */
const relativeBinPath = '../vitest.mjs'

const mochaPath = await import.meta.resolve('vitest')
const mochaPathDir = path.dirname(mochaPath)
const mochaBinPath = path.join(mochaPathDir, relativeBinPath)
await import(mochaBinPath)
