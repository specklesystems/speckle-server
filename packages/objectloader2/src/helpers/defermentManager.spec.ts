/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-explicit-any  */
import { describe, expect, test } from 'vitest'
import { DefermentManager } from './defermentManager.js'

describe('deferments', () => {
  test('defer one', async () => {
    const deferments = new DefermentManager({ maxSizeInMb: 1, ttlms: 1 })
    const x = deferments.defer({ id: 'id' })
    expect(x).toBeInstanceOf(Promise)
    deferments.undefer({ baseId: 'id', base: { id: 'id', speckle_type: 'type' } })
    const b = await x
    expect(b).toMatchSnapshot()
  })

  test('expireAt timeout', async () => {
    const now = 1
    const deferments = new DefermentManager({ maxSizeInMb: 1, ttlms: 1 })
    deferments['now'] = (): number => now
    const x = deferments.defer({ id: 'id' })
    expect(x).toBeInstanceOf(Promise)
    const d = deferments.get('id')
    expect(d).toBeDefined()
    expect(d?.getId()).toBe('id')
    expect((d as any).expiresAt).toBe(2)
    expect((d as any).ttl).toBe(1)
    expect((d as any).item).toBeUndefined()
    expect(d?.isExpired(1)).toBe(false)
    deferments.undefer({ baseId: 'id', base: { id: 'id', speckle_type: 'type' } })
    await x
    expect((d as any).expiresAt).toBe(2)
    expect((d as any).ttl).toBe(1)
    expect((d as any).item).toBeDefined()
    expect(d?.isExpired(1)).toBe(false)
    expect(d?.isExpired(3)).toBe(true)
  })
})
