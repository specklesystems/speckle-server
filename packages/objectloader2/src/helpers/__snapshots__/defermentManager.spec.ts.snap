// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deferments > defer one 1`] = `
{
  "id": "id",
  "speckle_type": "type",
}
`;
